docs/features/role-based-access-control/high-level-design.md

```diff
@@ -0,0 +1,100 @@
+
+This module, for now, will support checking permissions for the 3 resources subject to RBAC: organizations, projects and
+scenarios. For this, this module should provide 3 services, each of them implementing the following interface:
+
+
+`isOwner(resourceId: string, userId: string): boolean`
+
+Checks if the given user can perform the actions of a `Owner` for the provided resource id. This method should return 
+`true` if one of the following conditions apply:
+
+* the user is a platform admin;
+* the user has a role greater or equal to `Owner` for the resource being checked in the corresponding table (e.g. when 
+ checking if a user is owner of a project, `this.projectAclService.isOwner(projectId, userId)`, there must be a record 
+ in the `users_projects` table with role greater or equal to `Owner`).
+* the user has a role greater or equal to `Owner` for a resource hierarchically superior to the one being checked (e.g. 
+ again using projects as an example, if there is no record in the `users_projects` table, but there is a record for the 
+ given user with role greater or equal to `Owner` for the organization that owns the project).
+
+
+`isContributor(resourceId: string, userId: string): boolean`
+
+Checks if the given user can perform the actions of a `Contributor` for the provided resource id. This method should 
+return `true` if one of the following conditions apply:
+
+* the user is a platform admin;
+* the user has a role higher than `Contributor` for the resource being checked in the corresponding table (e.g. when 
+ checking if a user is contributor of a project, `this.projectAclService.isContributor(projectId, userId)`, there must 
+ be a record in the `users_projects` table for the role greater or equal to `Contributor`).
+* the user has the role higher than `Contributor` for a resource hierarchically superior to the one being checked (e.g. 
+ again using projects as an example, if there is no record in the `users_projects` table, but the user has a role higher 
+ or equal to `Contributor` for the organization that owns the project).
+
+
+`isViewer(resourceId: string, userId: string): boolean`
+
+Checks if the given user can perform the actions of a `Viewer` for the provided resource id. This method should return 
+`true` if one of the following conditions apply:
+
+* the user is a platform admin;
+* the user has a role higher than `Viewer` for the resource being checked in the corresponding table (e.g. when
+  checking if a user is contributor of a project, `this.projectAclService.isViewer(projectId, userId)`, there must
+  be a record in the `users_projects` table for the role greater or equal to `Viewer`).
+* the user has the role higher than `Viewer` for a resource hierarchically superior to the one being checked (e.g.
+  again using projects as an example, if there is no record in the `users_projects` table, but the user has a role higher
```

Comment:
Sometimes the particular description of implementation is a good-to-have but being too strict (pointing particular fields/tables) make it not being high-level document and denies the flexibility. I would leave the implementation details to low-level-design or not put them at all.

High level module API should be fair enough for HLD (i.e. how other modules will use this root lvl module)

Reply:
Though I agree with you that this document should be high-level, I think this description is important because it describes the hierarchical nature of the permissions.

Maybe I can also add some visuals explaining how other modules would use permissions, would that help?

Reply:
@ikas I am fine to keep the hierarchical nature of permissions (as long as we really want it) but move (or rather, remove) the technical details about particular entities.